### PR TITLE
First pass at parallelizing index creation

### PR DESCRIPTION
**Pull request by ee-usgs**
There are lots of sequential index build steps, but this just parallelizes indexes for Result and Activity.

### CODE REVIEW

File: src/main/java/gov/acwi/wqp/etl/activity/index/BuildActivityIndexes.java [hunk 1]
```diff
@@ -240,27 +244,28 @@ public Step buildActivityStationIdIndexStep() {
...
+					concurrent.makeFlow(buildActivityHuc6IndexStep()),
+					concurrent.makeFlow(buildActivityHuc8IndexStep()),
+					concurrent.makeFlow(buildActivityOrganizationIndexStep()),
+					concurrent.makeFlow(buildActivityProjectIdIndexStep()),
+					concurrent.makeFlow(buildActivitySampleMediaIndexStep()),
+					concurrent.makeFlow(buildActivitySiteIdIndexStep()),
+					concurrent.makeFlow(buildActivitySiteTypeIndexStep()),
+					concurrent.makeFlow(buildActivityStateIndexStep()),
+					concurrent.makeFlow(buildActivityStationIdIndexStep())
+				).build();
```
Comment (ee-usgs) [inline src/main/java/gov/acwi/wqp/etl/activity/index/BuildActivityIndexes.java:268]: Note that each step has been wrapped in a Flow.  'Flow' is the minimal unit that can be made to be parallel.

File: src/test/java/gov/acwi/wqp/etl/BaseFlowIT.java [hunk 2]
```diff
@@ -107,7 +107,7 @@
 	@Autowired
 	protected JobBuilderFactory jobBuilderFactory;
 	@Autowired
-	protected JobLauncherTestUtils jobLauncherTestUtils;
+	protected SmarterJobLauncherTestUtils jobLauncherTestUtils;
```
Comment (ee-usgs) [inline src/test/java/gov/acwi/wqp/etl/BaseFlowIT.java:110]: I just subclassed the JobLauncherTestUtils class and overroad the 'launchStep' method.  JobLauncherTestUtils is 'dumb' in that cannot find a Step if it is nested in a Flow.... and I just nested all the Steps in individual flows!  Dang!
See below for the implementation.

File: src/test/java/gov/acwi/wqp/etl/SmarterJobLauncherTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * The basic JobLauncherTestUtils cannot find a Step if it is nested within a Flow in a Flow, so this
+ * class provides a configurable backdoor to get it via reflection.
+ * @return
+ */
+@Component
+public class SmarterJobLauncherTestUtils extends JobLauncherTestUtils {
+
+	Object stepCreator;
```
Comment (ee-usgs) [inline src/test/java/gov/acwi/wqp/etl/SmarterJobLauncherTestUtils.java:20]: I added this field, which allows test classes to specify a 'creator' class that is used to create steps (e.g. BuildActivityIndex or BuildResultIndexes).  If the dumb implementation cannot find the step I want to individually execute for a test, hack it via reflection to hunt thru the stepCreator class to find a method of the same name as the step (all the access methods for building a step are named that way).

File: src/test/java/gov/acwi/wqp/etl/activity/index/BaseBuildActivityIndexesIT.java [hunk 4]
```diff
@@ -21,6 +26,8 @@ public void setUp() {
 				.start(buildActivityIndexesFlow)
 				.build()
 				.build();
+
+		jobLauncherTestUtils.setStepCreator(buildIndexes);
 		jobLauncherTestUtils.setJob(testJob);
 	}
 }
```
Comment (ee-usgs) [inline src/test/java/gov/acwi/wqp/etl/activity/index/BaseBuildActivityIndexesIT.java:33]: Now this small mod to the base class allows all the individual index tests for activity to work (20+ test classes for each table that has indexes... that I don't have to change now!)

File: src/test/java/gov/acwi/wqp/etl/result/index/BuildResultIndexesFlowIT.java [hunk 5]
```diff
@@ -3,15 +3,28 @@
...
 public class BuildResultIndexesFlowIT extends BaseBuildResultIndexesIT {
 
+	@Autowired
+	private ConfigurationService config;
+
+	@Test
+	public void indexConcurrencyShouldBeSetTo2FromTestProperties() {
+		assertEquals(2, config.getDbOperationConcurrency());
+	}
+
```
Comment (ee-usgs) [inline src/test/java/gov/acwi/wqp/etl/result/index/BuildResultIndexesFlowIT.java:27]: Just verifying that I can control concurrency via the env config property on line 17.

File: src/test/java/gov/acwi/wqp/etl/result/index/BuildResultActivityIndexIT.java [hunk 6]
```diff
@@ -3,15 +3,26 @@
...
 
+	@Autowired
+	private ConfigurationService config;
+
+	//This test really could be moved to a ConfigurationService test, but that test exists in a parallel merge request.
+	@Test
+	public void indexConcurrencyShouldBDefaultTo3() {
+		assertEquals(3, config.getDbOperationConcurrency());
+	}
+
```
Comment (ee-usgs) [inline src/test/java/gov/acwi/wqp/etl/result/index/BuildResultActivityIndexIT.java:25]: Confirming the default concurrency value.

**Review by ssoper-usgs — APPROVED** [review]

**Review by orosz-usgs — APPROVED** [review]

**Pull request merged**